File: grid/internal/webaccount/userid.go
package webaccount

import (
	"errors"
	"regexp"
	"strings"
)

// userid length bounds mirror the UserId schema in api/openapi.yaml.
const (
	minUseridLength = 3
	maxUseridLength = 32
)

var (
	// ErrInvalidDisplayName is returned when a display name cannot yield a
	// usable userid.
	ErrInvalidDisplayName = errors.New("display name must be two words and yield a 3-32 character userid")

	useridDisallowed = regexp.MustCompile(`[^a-z0-9'.]+`)
	useridCollapse   = regexp.MustCompile(`\.+`)
)

// DeriveUserid performs the authoritative derivation of an avatar login id from
// a display name. It mirrors deriveUserid in the website's src/lib/userid.js:
// lowercase, replace every run of non-[a-z0-9'.] characters with a single
// period, collapse consecutive periods, and trim leading/trailing periods.
func DeriveUserid(displayName string) string {
	lowered := strings.ToLower(displayName)
	replaced := useridDisallowed.ReplaceAllString(lowered, ".")
	collapsed := useridCollapse.ReplaceAllString(replaced, ".")
	return strings.Trim(collapsed, ".")
}

// ValidateDisplayName enforces the two-word rule and the derived userid length
// bounds, returning the derived userid on success.
func ValidateDisplayName(displayName string) (string, error) {
	words := strings.Fields(displayName)
	if len(words) != 2 {
		return "", ErrInvalidDisplayName
	}
	userid := DeriveUserid(displayName)
	if len(userid) < minUseridLength || len(userid) > maxUseridLength {
		return "", ErrInvalidDisplayName
	}
	return userid, nil
}

File: web/src/lib/userid.js
// Derives the avatar login id from a display name. The server performs the
// authoritative derivation; this mirror exists only for the live registration
// preview. Keep the two implementations in sync (see api/openapi.yaml UserId).
export function deriveUserid(displayName) {
  return String(displayName ?? "")
    .toLowerCase()
    .replace(/[^a-z0-9'.]+/g, ".") // non-allowed (incl. whitespace) -> period
    .replace(/\.+/g, ".") // collapse consecutive periods
    .replace(/^\.|\.$/g, ""); // trim leading/trailing period
}

export function displayNameWords(displayName) {
  return String(displayName ?? "")
    .trim()
    .split(/\s+/)
    .filter(Boolean);
}

// Returns an error message string when the display name is unusable, or null
// when it is valid.
export function validateDisplayName(displayName) {
  const words = displayNameWords(displayName);
  if (words.length !== 2) {
    return "Enter exactly two words: a first and last name.";
  }
  const userid = deriveUserid(displayName);
  if (userid.length < 3) {
    return "That name is too short to form a login ID.";
  }
  if (userid.length > 32) {
    return "That name is too long; please shorten it.";
  }
  return null;
}
